### DISCUSSION TITLE
latest/how_to/detect_and_annotate/

### DISCUSSION THREAD

**Discussion by giscus** [roboflow/supervision#873] [Docs]
# latest/how_to/detect_and_annotate/

A set of easy-to-use utilities that will come in handy in any computer vision project.

https://supervision.roboflow.com/latest/how_to/detect_and_annotate/

**Comment by artinm89** [comment]
this is so good but the annotating spend lot of space

**Reply by artinm89** [reply]
if there is any way for this issue pls let  me know

**Reply by SkalskiP** [reply]
Hi @artinm89 👋🏻 Did you faced any specific speed related issue?

**Comment by AnatolySamaris** [comment]
how do I annotate gray image as it has less dimension?

**Reply by SkalskiP** [reply]
Unfortunately, we do not support grayscale image annotations.

**Comment by Atef98** [comment]
How to annotate multiple labels like class_name, tracker_id, confidence, ....?

**Reply by SkalskiP** [reply]
Hi @Atef92! 👋🏻 Sorry for late response. Take a look here.

```python
labels = [
    f"#{tracker_id} {model.model.names[class_id]} {confidence:.2f}"
    for class_id, confidence, tracker_id
    in zip(detections.class_id, detections.confidence, detections.tracker_id)
]

annotated_image = label_annotator.annotate(
    scene=annotated_image, detections=detections, labels=labels)
```

**Comment by MubashirWaheed** [comment]
For some reason only bounding box annotation is working for me 

```js
from ultralytics import YOLO
import cv2 as cv 
import supervision as sv
coco_model = YOLO("yolov8n.pt")

image = cv.imread("./cars.png")
result = coco_model(image)[0]
# parse detections
detections = sv.Detections.from_ultralytics(result)
print('detections: ',detections)

detections = detections[detections.class_id == 2]
bounding_box_annotator = sv.MaskAnnotator()
label_annotator = sv.LabelAnnotator()

labels = [
    coco_model.model.names[class_id]
    for class_id
    in detections.class_id
]
annotated_image = bounding_box_annotator.annotate(scene=image, detections=detections)
annotated_image = label_annotator.annotate(scene=annotated_image, detections=detections, labels=labels)

sv.plot_image(annotated_image)

```
what is wrong in this code?

**Reply by LinasKo** [reply]
Hi @MubashirWaheed :wave:

`detections` you get from `from_ultralytics` is a universal object that can hold both object detections results (boxes) and segmentation results (masks). Some annotators only work on segmentation models.

Try changing to `yolov8n-seg.pt` and see what happens :slightly_smiling_face:

**Comment by luispatriciorf** [comment]
Hi guys,
How can I modify the label colors based on classes? I trained a model using yolov8 but I realized there were two classes that were the same like "player" and "players". When I use the annotation feature it assigns different colors which makes sense (they're two different classes) but both classes would have to have the same color.

I didn't find anything about that.

Thanks in advance!

Luis.

**Reply by LinasKo** [reply]
Hi @luispatriciorf,

Option 1 is to pass an argument `ColorPalette(colors=[Color(r=255, g=64, b=64), Color(r=255, g=161, b=160), ...])`, with N elements, for each of your N possible classes.

Option 2 is replacing the classes you have in detections, right after they're created. For example, if "player" and "players" are 1 and 8, you could do `detections.class_id[detections.class_id == 8] = 1`.

**Reply by SkalskiP** [reply]
To clarify Option 1 further: Every annotator in Supervision accepts a color argument. This can be a single [`sv.Color`](https://supervision.roboflow.com/latest/utils/draw/#supervision.draw.color.Color), in which case all detections will be marked identically, or an [`sv.ColorPalette`](https://supervision.roboflow.com/latest/utils/draw/#supervision.draw.color.ColorPalette) — a list of colors where each class is painted a different color. Referring to the example provided by Linus, if "player" and "players" are classes 1 and 8, you could define a custom `sv.ColorPalette` where the colors at indices 1 and 8 are the same and then pass it to your annotator. Here's an example:

```python
import supervision as sv

color_palette = sv.ColorPalette.from_hex([
    "A351FB",
    "FF4040", # <--- color with index 1
    "FFA1A0",
    "FF7633",
    "FFB633",
    "D1D435",
    "4CFB12",
    "94CF1A",
    "FF4040", # <--- color with index 8
    "1B9640",
])

bounding_box_annotator = sv.BoundingBoxAnnotator(color=color_palette)
```

**Reply by luispatriciorf** [reply]
Awesome! Thanks guys!

**Reply by luispatriciorf** [reply]
Guys,
It's me again, sorry. I don't know if it is a dumb question but how can I modify the ellipse diameter if I use one? Not the thickness but the diameter.

sv.EllipseAnnotator() only shows me thickness, start and end angle, and color.

Thanks!

**Reply by LinasKo** [reply]
It's alright - you have a problem you want to solve, so there's no dumb questions :)

Edit: use SkalskiP's method.

---

Indeed, we don't have a scale factor. The elipse size depends on the detection size.
To save time, here's a very scrappy workaround - it will depend on your use case, whether it's acceptable:

You can modify the `xyxy` values of the detections, right before annotation:
```
extra_diameter = 5
detections.xyxy[:, 0] -= extra_diameter
detections.xyxy[:, 1] -= extra_diameter
detections.xyxy[:, 2] += extra_diameter
detections.xyxy[:, 3] += extra_diameter
```

This should add additional pixels on each side. But again - this is very hacky and won't work for all use cases.
Does it solve yours?

**Reply by SkalskiP** [reply]
Or, you can use `pad_boxes` util.

```python
detections.xyxy = sv.pad_boxes(xyxy=detections.xyxy, px=5)
```

**Comment by pramudithaSJ** [comment]
This code work fine in my laptop , but in my raspberry pi 4 gives below error

labels = [ ^ TypeError: 'NoneType' object is not iterable

        while True:
            success, frame = self.cap.read()
            if success:
                results = self.yolo(frame)[0]
                detections = sv.Detections.from_ultralytics(results)
                detections = detections[detections.confidence > 0.70]

                if len(detections) > 0:
                    print(detections)
                    labels = [
                        class_name.split()[0]  # Extract just the numeric part
                        for class_name in detections['class_name']
                    ]
                    self.release_resources()
                    return jsonify({"message": "High confidence detections found","data":labels}), 200

**Reply by LinasKo** [reply]
Hi @pramudithaSJ :wave: 

Checking the code, `detections['class_name']` can return `None`. However, we do expect to see a names array in Ultralytics and create our own - it's strange. Could you please:

1. I assume this fails in `class_name.split()[0] for class_name in detections['class_name']`. Is that correct?
2. Verify you're using the same code, including the model number, the runtime (e.g. ultralytics, rather than `inference`)
3. Can you show me what's inside the detections object that fails? `print(detections)` should work for that.
4. Which model are you using? Ideally, seeing more code would be great.

**Reply by SkalskiP** [reply]
@pramudithaSJ, what version of `supervision` and `ultralytics` have you installed on your laptop and Raspberry Pi 4? What version of Python you have on both of those devices?

```bash
echo "Python version:"
python3 --version

# Check the supervision version
echo "supervision version:"
pip show supervision | grep Version

# Check the ultralytics version
echo "ultralytics version:"
pip show ultralytics | grep Version
```

**Comment by Guhan-C** [comment]
Can u do anything like this without libraries and using only open cv need to do ball trajectory finding

**Comment by Ilyosxon0000** [comment]
very insightful to know what have done!

**Comment by isidrosegui** [comment]
Is it possible to track a person and determine whether they have fallen and are lying on the floor? If so, what would be the recommended approach for detecting and tracking this condition in real time?
